src/ipnet.rs

```diff
@@ -0,0 +1,828 @@
+        u32::MAX << (IPV4_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return true iff the subnet contains only the base address i.e. the
+    /// size is exactly one address.
+    pub fn is_host_net(&self) -> bool {
+        self.width == IPV4_NET_WIDTH_MAX
+    }
+
+    /// Return the number of addresses contained within this subnet.
+    pub fn size(&self) -> u32 {
+        1u32 << (IPV4_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return the prefix address (the base address with the mask applied).
+    pub fn prefix(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr & self.mask())
+    }
+
+    /// Return the network address for subnets wider than a /31 or /32.
+    pub fn network(&self) -> Option<Ipv4Addr> {
+        (self.width < 31).then(|| {
+            let addr: u32 = self.addr.into();
+            Ipv4Addr::from(addr & self.mask())
+        })
+    }
+    /// Return the broadcast address for subnets wider than a /31 or /32.
+    pub fn broadcast(&self) -> Option<Ipv4Addr> {
+        (self.width < 31).then(|| {
+            let addr: u32 = self.addr.into();
+            Ipv4Addr::from(addr | !self.mask())
+        })
+    }
+
+    /// Return the first address within this subnet.
+    pub fn first_addr(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr & self.mask())
+    }
+
+    /// Return the last address within this subnet.
+    pub fn last_addr(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr | !self.mask())
+    }
+
+    /// Return the first host address within this subnet; for subnets wider
+    /// than a /31 or /32 that is the first address after the network address.
+    pub fn first_host(&self) -> Ipv4Addr {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let first = addr & mask;
+        if self.width == 31 || self.width == 32 {
+            Ipv4Addr::from(first)
+        } else {
+            Ipv4Addr::from(first + 1)
+        }
+    }
+
+    /// Return the last host address within this subnet; for subnets wider
+    /// than a /31 or /32 that is the address immediately before the broadcast
+    /// address.
+    pub fn last_host(&self) -> Ipv4Addr {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let last = addr | !mask;
+        if self.width == 31 || self.width == 32 {
+            Ipv4Addr::from(last)
+        } else {
+            Ipv4Addr::from(last - 1)
+        }
+    }
+
+    /// Return `true` iff the given IP address is within the subnet.
+    pub fn contains(&self, other: Ipv4Addr) -> bool {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let other: u32 = other.into();
+
+        (addr & mask) == (other & mask)
+    }
+
+    /// Produce an Interator over all addresses within this subnet.
+    pub fn addr_iter(&self) -> impl Iterator<Item = Ipv4Addr> {
```

Comment:
Is it worth making this a newtype, to avoid someone passing in the return value of `addr_iter()` when they meant `host_iter()` or vice-versa?

Reply:
meh? So we'd have `struct Host(Ipv4Addr)` or something? I don't know; what do you think?

Reply:
Yeah, let's just punt. We can do it later if we need to.